Client/Frontend/Browser/FaviconHandler.swift

```diff
@@ -0,0 +1,86 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import Storage
+import SDWebImage
+import Deferred
+
+class FaviconHandler {
+    static let MaximumFaviconSize = 1 * 1024 * 1024 // 1 MiB file size limit
+
+    private var tabObservers: TabObservers!
+
+    init() {
+        self.tabObservers = registerFor(.didLoadPageMetadata, queue: .main)
+    }
+
+    deinit {
+        unregister(tabObservers)
+    }
+
+    func loadFaviconURL(_ faviconURL: String, forTab tab: Tab) -> Deferred<Maybe<(Favicon, Data?)>> {
+        guard let iconURL = URL(string: faviconURL), let currentURL = tab.url else {
+            return deferMaybe(FaviconError())
+        }
+
+        let deferred = Deferred<Maybe<(Favicon, Data?)>>()
+        let manager = SDWebImageManager.shared()
+        let url = currentURL.absoluteString
+        let site = Site(url: url, title: "")
+        let options: SDWebImageOptions = tab.isPrivate ? SDWebImageOptions([.lowPriority, .cacheMemoryOnly]) : SDWebImageOptions([.lowPriority])
+
+        var fetch: SDWebImageOperation? = nil
+
+        let onProgress: SDWebImageDownloaderProgressBlock = { (receivedSize, expectedSize, _) -> Void in
+            if receivedSize > FaviconHandler.MaximumFaviconSize || expectedSize > FaviconHandler.MaximumFaviconSize {
+                fetch?.cancel()
+            }
+        }
+
+        let onCompleted: SDInternalCompletionBlock = { (img, data, _, _, _, url) -> Void in
+            guard let img = img, let urlString = url?.absoluteString else {
+                deferred.fill(Maybe(failure: FaviconError()))
+                return
+            }
+
+            let favicon = Favicon(url: urlString, date: Date())
+            favicon.width = Int(img.size.width)
+            favicon.height = Int(img.size.height)
+
+            tab.favicons.append(favicon)
+
+            if !tab.isPrivate, let appDelegate = UIApplication.shared.delegate as? AppDelegate, let profile = appDelegate.profile {
+                profile.favicons.addFavicon(favicon, forSite: site).upon { _ in
+                    deferred.fill(Maybe(success: (favicon, data)))
+                }
+            } else {
+                deferred.fill(Maybe(success: (favicon, data)))
+            }
+        }
+
+        fetch = manager.loadImage(with: iconURL, options: options, progress: onProgress, completed: onCompleted)
+        return deferred
+    }
+}
+
+extension FaviconHandler: TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {
+        tab.favicons.removeAll(keepingCapacity: false)
+        if let faviconURL = metadata.faviconURL {
+            loadFaviconURL(faviconURL, forTab: tab).uponQueue(DispatchQueue.main) { result in
+                if let (favicon, data) = result.successValue {
+                    TabEvent.post(.didLoadFavicon(favicon, with: data), for: tab)
```

Comment:
You can post from any queue, the handler chooses which queue to be notified on. You've explicitly given `queue: .main` when registering `TopTabsViewController` below.
 
Additionally, I think you can replace `upon` with `>>==` to eliminate the `if let _ = result.successValue`.

```swift
loadFaviconURL(faviconURL, forTab: tab) >>== { (favicon, data) in 
    TabEvent.post(.didLoadFavicon(favicon, with: data), for: tab)
}
```